pageserver/src/tenant/timeline.rs

```diff
@@ -4669,28 +4679,60 @@ impl Timeline {
         // image layers  <100000000..100000099> and <200000000..200000199> are not completely covering it.
         let mut start = Key::MIN;
 
-        let check_for_image_layers = if let LastImageLayerCreationStatus::Incomplete = last_status {
-            info!(
-                "resuming image layer creation: last_status={:?}",
-                last_status
-            );
-            true
-        } else {
-            self.should_check_if_image_layers_required(lsn)
-        };
+        let check_for_image_layers =
+            if let LastImageLayerCreationStatus::Incomplete { last_key } = last_status {
+                info!(
+                    "resuming image layer creation: last_status=incomplete, continue from {}",
+                    last_key
+                );
+                true
+            } else {
+                self.should_check_if_image_layers_required(lsn)
+            };
 
         let mut batch_image_writer = BatchLayerWriter::new(self.conf).await?;
 
         let mut all_generated = true;
 
         let mut partition_processed = 0;
         let total_partitions = partitioning.parts.len();
+        let mut last_partition_processed = None;
+        let mut partition_parts = partitioning.parts.clone();
+
+        if let LastImageLayerCreationStatus::Incomplete { last_key } = last_status {
+            // We need to skip the partitions that have already been processed.
+            for (i, partition) in partition_parts.iter().enumerate() {
+                if last_key <= partition.end().unwrap() {
+                    // ```plain
+                    // |------|--------|----------|------|
+                    //              ^last_key
+                    //                    ^start from this partition
+                    // ```
+                    // Why `i+1` instead of `i`?
+                    // It is possible that the user did some writes after the previous image layer creation attempt so that
+                    // a relation grows in size, and the last_key is now in the middle of the partition. In this case, we
+                    // still want to skip this partition, so that we can make progress and avoid generating image layers over
+                    // the same partition.
+                    partition_parts.rotate_left((i + 1) % total_partitions);
+                    // Update the start key to the partition start.
+                    start = partition_parts[0].start().unwrap();
```

Comment:
Do we need to wrap around, or can we simply end when we reach the last partition? In some sense, we're just finishing a compaction we already started.

There's two reasons why I think this might be better:

1. If we wrap around we're likely compacting a keyspace that was recently compacted anyway, and we can end up spinning (is this accurate?).

2. It might introduce subtle bugs if the code doesn't expect it.

Reply:
1. yes; 2. I'm overall confident that the code can handle the case well but I agree to remove this wrap around.

Reply:
removed in https://github.com/neondatabase/neon/pull/10660/commits/91032398a748eaa79013dfb0491ad593d37c677f